src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/DefaultPageApplicationModelProvider.cs

```diff
@@ -237,12 +258,19 @@ protected virtual PagePropertyModel CreatePropertyModel(PropertyInfo property)
             }
 
             var propertyAttributes = property.GetCustomAttributes(inherit: true);
-            var handlerAttributes = property.DeclaringType.GetCustomAttributes(inherit: true);
 
-            // Look for binding info on the handler if nothing is specified on the property.
-            // This allows BindProperty attributes on handlers to apply to properties.
-            var bindingInfo = BindingInfo.GetBindingInfo(propertyAttributes) ??
-                BindingInfo.GetBindingInfo(handlerAttributes);
+            var propertyMetadata = _modelMetadataProvider.GetMetadataForProperty(property.DeclaringType, property.Name);
+            var bindingInfo = BindingInfo.GetBindingInfo(propertyAttributes, propertyMetadata);
+
+            if (bindingInfo == null)
+            {
+                // Look for binding info on the handler if nothing is specified on the property.
+                // This allows BindProperty attributes on handlers to apply to properties.
+                var handlerType = property.DeclaringType;
+                var handlerAttributes = handlerType.GetCustomAttributes(inherit: true);
+                var handlerMetadata = _modelMetadataProvider.GetMetadataForType(property.DeclaringType);
+                bindingInfo = BindingInfo.GetBindingInfo(handlerAttributes, handlerMetadata);
+            }
```

Comment:
Not sure I've seen this code before. Why doesn't this merge the property and handler attributes?

Even if these must be kept separate, why not use `propertyMetadata` in this block?

Reply:
> Even if these must be kept separate, why not use propertyMetadata in this block?

I don't follow. We're basically transferring any handler specific metadata to the property if the property has nothing defined.